pyop2/types/set.py

```diff
@@ -543,13 +537,15 @@ def __init__(self, sets):
         assert all(s is None or isinstance(s, GlobalSet) or ((s.layers == self._sets[0].layers).all() if s.layers is not None else True) for s in sets), \
             "All components of a MixedSet must have the same number of layers."
         # TODO: do all sets need the same communicator?
-        self.comm = mpi.internal_comm(functools.reduce(lambda a, b: a or b, map(lambda s: s if s is None else s.comm, sets)))
+        self.comm = mpi.internal_comm(
+            functools.reduce(
+                lambda a, b: a or b,
+                map(lambda s: s if s is None else s.comm, sets)
+            ),
```

Comment:
```suggestion
            pytools.single_valued(s.comm for s in sets if s is not None)
```

Reply:
Do I need to add pytools as a hard dependency to do this?

Reply:
Or `val, = set(s.comm for s in sets if s is not None)` if you don't want pytools.

Reply:
Nice idea, but `mpi4py.MPI.Intracomm` is not a hashable type, so using a Python `set()` won't work here.

Reply:
We always get pytools with loopy, and writing a replacement would be trivial if we eventually rip loopy out, so I see no issue in using it here. It should be listed as a formal requirement though.

Reply:
Yes, this is done :slightly_smiling_face: